service/matching/db.go

```diff
@@ -167,54 +171,61 @@ func (db *taskQueueDB) takeOverTaskQueueLocked(
 	})
 	switch err.(type) {
 	case nil:
-		response.TaskQueueInfo.Kind = db.queue.Partition().Kind()
-		response.TaskQueueInfo.ExpiryTime = db.expiryTime()
-		response.TaskQueueInfo.LastUpdateTime = timestamp.TimeNowPtrUtc()
-		if _, err := db.store.UpdateTaskQueue(ctx, &persistence.UpdateTaskQueueRequest{
-			RangeID:       response.RangeID + 1,
-			TaskQueueInfo: response.TaskQueueInfo,
-			PrevRangeID:   response.RangeID,
-		}); err != nil {
-			return err
-		}
-		db.rangeID = response.RangeID + 1
+		db.rangeID = response.RangeID
 		db.subqueues = db.ensureDefaultSubqueuesLocked(
 			response.TaskQueueInfo.Subqueues,
 			response.TaskQueueInfo.AckLevel,
 			response.TaskQueueInfo.ApproximateBacklogCount,
 		)
+		err := db.updateTaskQueueLocked(ctx, true)
+		if err != nil {
+			db.rangeID = 0
+			return err
+		}
+		// We took over the task queue and are not sure what tasks may have been written
+		// before. Set max read level of all subqueues to just before our new block.
+		maxReadLevel := rangeIDToTaskIDBlock(db.rangeID, db.config.RangeSize).start - 1
+		for _, s := range db.subqueues {
+			s.maxReadLevel = maxReadLevel
+		}
 		return nil
 
 	case *serviceerror.NotFound:
 		db.rangeID = initialRangeID
 		db.subqueues = db.ensureDefaultSubqueuesLocked(nil, 0, 0)
 		if _, err := db.store.CreateTaskQueue(ctx, &persistence.CreateTaskQueueRequest{
-			RangeID:       initialRangeID,
+			RangeID:       db.rangeID,
 			TaskQueueInfo: db.cachedQueueInfo(),
 		}); err != nil {
 			db.rangeID = 0
 			return err
 		}
+		db.lastWrite = time.Now()
+		// In this case, ensureDefaultSubqueuesLocked already initialized subqueue 0 to have
+		// ackLevel and maxReadLevel 0, so we don't need to initialize them.
+		bugIf(db.subqueues[0].maxReadLevel != 0, "bug: should have maxReadLevel 0 here")
+		bugIf(db.subqueues[0].AckLevel != 0, "bug: should have ackLevel 0 here")
 		return nil
 
 	default:
 		return err
 	}
 }
 
-func (db *taskQueueDB) renewTaskQueueLocked(
-	ctx context.Context,
-	rangeID int64,
-) error {
+func (db *taskQueueDB) updateTaskQueueLocked(ctx context.Context, incrementRangeId bool) error {
+	newRangeID := db.rangeID
+	if incrementRangeId {
+		newRangeID++
+	}
 	if _, err := db.store.UpdateTaskQueue(ctx, &persistence.UpdateTaskQueueRequest{
-		RangeID:       rangeID,
+		RangeID:       newRangeID,
 		TaskQueueInfo: db.cachedQueueInfo(),
```

Comment:
Q: What part of `cachedQueueInfo` is "cached"? Or does that mean "local" / "previously read"?

Reply:
Think write-back cache, not write-through/read-through. Updated in memory but maybe not persisted yet.